File: frontend/src/services/SectionService.ts
import api from './api';
import type { Section } from '../types/Section';

class SectionService {
    // Get all sections
    async getAllSections(): Promise<Section[]> {
        const response = await api.get<Section[]>('/sections');
        return response.data;
    }

    // Get all published sections
    async getPublishedSections(): Promise<Section[]> {
        const response = await api.get<Section[]>('/sections/published');
        return response.data;
    }

    // Get all sections ordered by display order
    async getAllSectionsOrdered(): Promise<Section[]> {
        const response = await api.get<Section[]>('/sections/ordered');
        return response.data;
    }

    // Get a single section by ID
    async getSectionById(id: number): Promise<Section> {
        const response = await api.get<Section>(`/sections/${id}`);
        return response.data;
    }

    // Get a section by slug
    async getSectionBySlug(slug: string): Promise<Section> {
        const response = await api.get<Section>(`/sections/slug/${slug}`);
        return response.data;
    }

    // Create a new section
    async createSection(section: Section): Promise<Section> {
        const response = await api.post<Section>('/sections', section);
        return response.data;
    }

    // Create a section with details
    async createSectionWithDetails(
        slug: string,
        title: string,
        content: string,
        description?: string,
        displayOrder?: number,
        published: boolean = true,
        componentType?: string
    ): Promise<Section> {
        const params = new URLSearchParams();
        params.append('slug', slug);
        params.append('title', title);
        params.append('content', content);
        if (description) params.append('description', description);
        if (displayOrder !== undefined) params.append('displayOrder', displayOrder.toString());
        params.append('published', published.toString());
        if (componentType) params.append('componentType', componentType);

        const response = await api.post<Section>(`/sections/create?${params.toString()}`);
        return response.data;
    }

    // Update an existing section
    async updateSection(id: number, section: Section): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}`, section);
        return response.data;
    }

    // Update section with details
    async updateSectionWithDetails(
        id: number,
        slug?: string,
        title?: string,
        description?: string,
        content?: string,
        displayOrder?: number,
        published?: boolean,
        componentType?: string
    ): Promise<Section> {
        const params = new URLSearchParams();
        if (slug) params.append('slug', slug);
        if (title) params.append('title', title);
        if (description) params.append('description', description);
        if (content) params.append('content', content);
        if (displayOrder !== undefined) params.append('displayOrder', displayOrder.toString());
        if (published !== undefined) params.append('published', published.toString());
        if (componentType) params.append('componentType', componentType);

        const response = await api.put<Section>(`/sections/${id}/update?${params.toString()}`);
        return response.data;
    }

    // Delete a section
    async deleteSection(id: number): Promise<void> {
        await api.delete(`/sections/${id}`);
    }

    // Publish a section
    async publishSection(id: number): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/publish`);
        return response.data;
    }

    // Unpublish a section
    async unpublishSection(id: number): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/unpublish`);
        return response.data;
    }

    // Update display order
    async updateDisplayOrder(id: number, newOrder: number): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/order?newOrder=${newOrder}`);
        return response.data;
    }

    // Reorder multiple sections
    async reorderSections(sectionIds: number[]): Promise<Section[]> {
        const response = await api.put<Section[]>('/sections/reorder', sectionIds);
        return response.data;
    }

    // Update section content
    async updateContent(id: number, content: string): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/content?content=${encodeURIComponent(content)}`);
        return response.data;
    }

    // Update section title
    async updateTitle(id: number, title: string): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/title?title=${encodeURIComponent(title)}`);
        return response.data;
    }

    // Update section description
    async updateDescription(id: number, description: string): Promise<Section> {
        const response = await api.put<Section>(`/sections/${id}/description?description=${encodeURIComponent(description)}`);
        return response.data;
    }
}

export default new SectionService();

File: backend/src/main/java/com/exo/service/SectionService.java
package com.exo.service;

import com.exo.model.Section;
import com.exo.repository.SectionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class SectionService {

    @Autowired
    private SectionRepository sectionRepository;

    /* ==========================
     *      BASIC CRUD
     * ==========================
     */

    public Section saveSection(Section section) {
        return sectionRepository.save(section);
    }

    public Optional<Section> findById(Long id) {
        return sectionRepository.findById(id);
    }

    public List<Section> findAll() {
        return sectionRepository.findAll();
    }

    public void deleteById(Long id) {
        sectionRepository.deleteById(id);
    }

    /* ==========================
     *      SECTION MANAGEMENT
     * ==========================
     */

    public Section createSection(String slug, String title, String description, 
                                String content, Integer displayOrder, Boolean published, String componentType) {
        Section section = new Section();
        section.setSlug(slug);
        section.setTitle(title);
        section.setDescription(description);
        section.setContent(content);
        section.setDisplayOrder(displayOrder != null ? displayOrder : 0);
        section.setPublished(published != null ? published : true);
        section.setComponentType(componentType);
        return sectionRepository.save(section);
    }

    public Section updateSection(Long id, String slug, String title, String description,
                                String content, Integer displayOrder, Boolean published, String componentType) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            if (slug != null) section.setSlug(slug);
            if (title != null) section.setTitle(title);
            if (description != null) section.setDescription(description);
            if (content != null) section.setContent(content);
            if (displayOrder != null) section.setDisplayOrder(displayOrder);
            if (published != null) section.setPublished(published);
            if (componentType != null) section.setComponentType(componentType);
            return sectionRepository.save(section);
        }
        return null;
    }

    /* ==========================
     *      PUBLICATION MANAGEMENT
     * ==========================
     */

    public Section publishSection(Long id) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setPublished(true);
            return sectionRepository.save(section);
        }
        return null;
    }

    public Section unpublishSection(Long id) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setPublished(false);
            return sectionRepository.save(section);
        }
        return null;
    }

    /* ==========================
     *      SEARCH & FILTER
     * ==========================
     */

    public Section findBySlug(String slug) {
        return sectionRepository.findBySlug(slug);
    }

    public List<Section> getPublishedSections() {
        return sectionRepository.findByPublishedTrueOrderByDisplayOrderAsc();
    }

    public List<Section> getDraftSections() {
        List<Section> allSections = sectionRepository.findAll();
        return allSections.stream()
                .filter(section -> !section.getPublished())
                .toList();
    }

    public List<Section> getAllSectionsOrdered() {
        List<Section> allSections = sectionRepository.findAll();
        return allSections.stream()
                .sorted((s1, s2) -> {
                    int order1 = s1.getDisplayOrder() != null ? s1.getDisplayOrder() : 0;
                    int order2 = s2.getDisplayOrder() != null ? s2.getDisplayOrder() : 0;
                    return Integer.compare(order1, order2);
                })
                .toList();
    }

    /* ==========================
     *      ORDERING MANAGEMENT
     * ==========================
     */

    public Section updateDisplayOrder(Long id, Integer newOrder) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setDisplayOrder(newOrder);
            return sectionRepository.save(section);
        }
        return null;
    }

    public List<Section> reorderSections(List<Long> sectionIds) {
        List<Section> sections = sectionRepository.findAllById(sectionIds);
        for (int i = 0; i < sections.size(); i++) {
            sections.get(i).setDisplayOrder(i);
        }
        return sectionRepository.saveAll(sections);
    }

    /* ==========================
     *      CONTENT MANAGEMENT
     * ==========================
     */

    public Section updateContent(Long id, String content) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setContent(content);
            return sectionRepository.save(section);
        }
        return null;
    }

    public Section updateTitle(Long id, String title) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setTitle(title);
            return sectionRepository.save(section);
        }
        return null;
    }

    public Section updateDescription(Long id, String description) {
        Optional<Section> optional = sectionRepository.findById(id);
        if (optional.isPresent()) {
            Section section = optional.get();
            section.setDescription(description);
            return sectionRepository.save(section);
        }
        return null;
    }
}
